src/main/java/org/jenkinsci/plugins/registry/notification/events/EventType.java

```diff
@@ -0,0 +1,64 @@
+package org.jenkinsci.plugins.registry.notification.events;
+
+import hudson.EnvVars;
+import hudson.model.AbstractDescribableImpl;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.registry.notification.webhook.PushNotification;
+import org.jenkinsci.plugins.registry.notification.webhook.WebHookPayload;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import javax.annotation.Nonnull;
+import java.util.logging.Logger;
+
+public abstract class EventType extends AbstractDescribableImpl<EventType> {
+    private static final Logger logger = Logger.getLogger(EventType.class.getName());
+    public static final String KEY_DOCKER_IMAGE_DIGEST = WebHookPayload.PREFIX + "DOCKER_IMAGE_DIGEST";
+
+    public boolean accepts(@Nonnull String jsonTypeCode ){
+        return jsonTypeCode.equals(getType());
+    }
+
+    public void buildEnvironment(@Nonnull EnvVars envs, @Nonnull PushNotification push) {
+        // Each event type can have a unique time stamp field name
+        WebHookPayload payload = push.getWebHookPayload();
+        if( payload != null ){
+            JSONObject data = payload.getData();
+            if ( data != null) {
+                JSONObject contents = data.getJSONObject("contents");
+                if (contents != null) {
+                    DateTimeFormatter parser = ISODateTimeFormat.dateTimeParser();
+                    push.setPushedAt(parser.parseDateTime(getTimeStamp(contents)).toDate());
+
+                    envs.put(EventTypeDescriptor.ENVIRONMENT_KEY, getType());
+                    if (hasDigest()) {
+                        envs.put(KEY_DOCKER_IMAGE_DIGEST, contents.getString("digest"));
```

Comment:
_nit_ `getDigest` similar to `getTimestamp` for consistency.

Reply:
I used has because it's a boolean (does the type have one or not)

Reply:
> (does the type have one or not)

Unsure what you mean by that?

Reply:
Clearer wording might be "does this type (the child class) have a digest or not" . Some event types, like SecurityScanCompleted, do not have a digest in the JSON payload

hasDigest() reads better to me than isDigest() for this boolean ("is" being the usual boolean accessor verb).  Do you have other ideas? isDigestPresent? isDigestValidForType?  isDigestSomethingThisTypeOfWebHookReturns? ;)

Reply:
`hasDigest` is fine, though probably not needed if you have.

```java
String digest = getDigest(contents);
if (digest != null) {
    envs.put(KEY_DOCKER_IMAGE_DIGEST, digest);
}
```

Reply:
Right.   That would push the logic of contents.getString("digest") back to DockerTrustedRegistryWebHookPayload() - maybe that's cleaner anyway.

Reply:
... if it's not in the JSON, then the setDigest call with just set the null